macaroons/constraints.go

```diff
@@ -0,0 +1,86 @@
+package macaroons
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"gopkg.in/macaroon-bakery.v1/bakery/checkers"
+	macaroon "gopkg.in/macaroon.v1"
+)
+
+// Constraint type adds a layer of indirection over macaroon caveats and
+// checkers.
+type Constraint func(*macaroon.Macaroon)
+
+// AddConstraints returns new derived macaroon by applying every passed
+// constraint and tightening its restrictions.
+func AddConstraints(mac *macaroon.Macaroon, cs ...Constraint) *macaroon.Macaroon {
+	newMac := mac.Clone()
+	for _, constraint := range cs {
+		constraint(newMac)
+	}
+	return newMac
+}
+
+// Each *Constraint function is a functional option, which takes a pointer
+// to the macaroon and adds another restriction to it. For each *Constraint,
+// the corresponding *Checker is provided.
+
+// AllowConstraint restricts allowed operations set to the ones
+// passed to it.
+func AllowConstraint(ops ...string) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		caveat := checkers.AllowCaveat(ops...)
+		mac.AddFirstPartyCaveat(caveat.Condition)
+	}
+}
+
+// AllowChecker wraps default checkers.OperationChecker.
+func AllowChecker(method string) checkers.Checker {
+	return checkers.OperationChecker(method)
+}
+
+// TimeoutConstraint restricts the lifetime of the macaroon
+// to the amount of seconds given.
+func TimeoutConstraint(seconds int64) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		macaroonTimeout := time.Duration(seconds)
+		requestTimeout := time.Now().Add(time.Second * macaroonTimeout)
+		caveat := checkers.TimeBeforeCaveat(requestTimeout)
+		mac.AddFirstPartyCaveat(caveat.Condition)
+	}
+}
+
+// TimeoutChecker wraps default checkers.TimeBefore checker.
+func TimeoutChecker() checkers.Checker {
+	return checkers.TimeBefore
+}
+
+// IPLockConstraint locks macaroon to a specific IP address.
+// If address is an empty string, this constraint does nothing to
+// accommodate default value's desired behavior.
+func IPLockConstraint(ipAddr string) func(*macaroon.Macaroon) {
+	return func(mac *macaroon.Macaroon) {
+		if ipAddr != "" {
```

Comment:
This checks whether `ipAddr`, received from `-macaroonip` option, is not a default `""` value, and if it isn't, adds another caveat.